### PR TITLE
Feature/3dvar dumprst

### PR DESCRIPTION
**Pull request by guillaumevernieres**

### CODE REVIEW

**Comment by travissluka** [comment]
I'll go ahead and be the jerk who demands non-blank pull request descriptions in the future.

File: src/Fields/Fields.cc [hunk 1]
```diff
@@ -217,14 +217,14 @@ namespace soca {
     int ncat = -1;
     int nf = -1;
     soca_field_sizes_f90(keyFlds_, nx, ny, nzo, nzi, ncat, nf);
-    // os << std::endl << "  Resolution = " << nx << ", " << ny
-    //   << ", Fields = " << nf;
+    os << std::endl << "  Resolution = " << nx << ", " << ny
```
Comment (aerorahul) [inline src/Fields/Fields.cc]: I don't think this is allowed.
Use Log::debug instead or other ways of displaying information, unless this was part of debugging, in which case, needs clean-up.

File: src/Model/Model.cc [hunk 2]
```diff
@@ -15,7 +15,7 @@
 #include "oops/util/Logger.h"
 
 #include "src/Fields/Fields.h"
-#include "src/Fortran.h"
+#include "ModelFortran.h"
```
Comment (aerorahul) [inline src/Model/Model.cc:18]: Why breaking from convention here?

File: src/Model/soca_model_mod.F90 [hunk 3]
```diff
@@ -104,17 +127,21 @@ subroutine soca_propagate(self, flds, fldsdate)
...
+       call step_MOM(self%mom6_config%forces, &
+                     &self%mom6_config%fluxes, &
+                     &self%mom6_config%sfc_state, &
+                     &self%mom6_config%Time, &
+                     &real(self%mom6_config%dt_forcing, kind=8), &
+                     &self%mom6_config%MOM_CSp,&
+                     &start_cycle=.false.,&
+                     &cycle_length=self%mom6_config%MOM_CSp%dt)
+    else
+       !TODO: Read file
```
Comment (aerorahul) [inline src/Model/soca_model_mod.F90]: Is this valid?
An empty else block?

File: src/Model/soca_model.interface.F90 [hunk 4]
```diff
@@ -52,12 +52,7 @@ subroutine c_soca_setup(c_confspec, c_key_geom, c_key_model) bind (c,name='soca_
   model%advance_mom6 = config_get_int(c_confspec,"advance_mom6")
 
   ! Initialize mom6
-  if (model%advance_mom6==1) then
-     call soca_create(model)
-     model%integration_initialized = .true.
-  else
-     print *,"Not initializing MOM6"
-  end if
+  call soca_setup(model)
```
Comment (aerorahul) [inline src/Model/soca_model.interface.F90:55]: Ooh, I like this!

File: src/Model/soca_model.interface.F90 [hunk 5]
```diff
@@ -76,10 +71,7 @@ subroutine c_soca_delete(c_key_conf) bind (c,name='soca_delete_f90')
   type(soca_model), pointer :: model
 
   call soca_model_registry%get(c_key_conf, model)
-  if (model%advance_mom6==1) then
-     call soca_delete(model)
-     model%integration_initialized = .false.     
-  end if  
+  call soca_delete(model)
```
Comment (aerorahul) [inline src/Model/soca_model.interface.F90:73]: and this.

File: src/Model/soca_model.interface.F90 [hunk 6]
```diff
@@ -109,9 +101,39 @@ subroutine c_soca_prepare_integration(c_key_model, c_key_state) &
   call soca_field_registry%get(c_key_state,flds)
   call soca_model_registry%get(c_key_model, model)
 
+  call soca_prepare_integration(model, flds)
```
Comment (aerorahul) [inline src/Model/soca_model.interface.F90]: much cleaner.
Should this be soca_initialize_integration to correspond with soca_finalize_integration?

**Review by aerorahul — COMMENTED** [review]
First glance comments.
Will run tests now.

File: src/Model/Model.cc (hunk 2 above)
Comment (guillaumevernieres) [reply src/Model/Model.cc:18]: cleaning up the Fortran.h that's in src

File: src/Model/soca_model_mod.F90 (hunk 3 above)
Comment (guillaumevernieres) [reply src/Model/soca_model_mod.F90]: I wondered too, gcc didn't complain. Will try with intel.

File: src/Model/soca_model.interface.F90 (hunk 6 above)
Comment (guillaumevernieres) [reply src/Model/soca_model.interface.F90]: Yes, left over from copy paste of qg ...

**Comment by guillaumevernieres** [comment]
@travissluka : Sorry, I forgot:
Addition of an application to checkpoint the model, which cannot be done anymore when running the 3dvar. Our JEDI friends decided to not call the model in the 3dvar cost function anymore. So while a model object is still being constructed, there are no calls to init and finalize (and currently the only way to checkpoint with the analysis in the state). There are other ways to do this, like having state.save() actually dump a full mom6 restart ... But that's not how it is currently done in soca.

File: src/Model/Model.cc (hunk 2 above)
Comment (guillaumevernieres) [reply src/Model/Model.cc:18]: @aerorahul : Do you mean just sticking to Fortran.h?

File: src/Model/soca_model_mod.F90 (hunk 3 above)
Comment (guillaumevernieres) [reply src/Model/soca_model_mod.F90]: @aerorahul : Added log comment to fill the block.

File: src/Model/soca_model_mod.F90 [hunk 7]
```diff
@@ -104,17 +127,23 @@ subroutine soca_propagate(self, flds, fldsdate)
...
+       WRITE(buf,*) 'IO Advance of MOM6: NOT IMPLEMENTED'
```
Comment (travissluka) [inline src/Model/soca_model_mod.F90]: write and print statements that are executed on all cores need to go away, eventually . I'm already getting swamped with repetitive log lines on my 8 core machine, and dread the day I do a 720 core run on gaea.

File: mains/CheckpointModel.h [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+#include "oops/runs/Application.h"
+#include "oops/util/DateTime.h"
+#include "oops/util/Duration.h"
+#include "oops/util/Logger.h"
+
+namespace soca {
+
+  class CheckpointModel : public oops::Application {
+   public:
+    static const std::string classname() {return "soca::ErrorCovariance";}
```
Comment (travissluka) [inline mains/CheckpointModel.h]: soca::ErrorCovariance ?

**Comment by guillaumevernieres** [comment]
I'm finding quite a few issues when building with intel17, I'm closing the PR and will resubmit later today. Hopefully.

**Comment by guillaumevernieres** [comment]
@travissluka , @aerorahul :  The intel fix will be a different pr, since the problem also exist in develop.

File: test/testinput/checkpointmodel.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  read_from_file: 1
+  date: 2018-04-15T00:00:00Z
+  basename: ./INPUT/
+  ocn_filename: MOM.res.nc
+  ice_filename: ice_model.res.nc
+analysis:
+  read_from_file: 1
+  date: 2018-04-15T00:00:00Z
+  basename: ./Data/
+  ocn_filename: ocn.3dvar.an.2018-04-15T00:00:00Z.nc
```
Comment (travissluka) [inline test/testinput/checkpointmodel.yml]:
```suggestion
  ocn_filename: ocn.3dvarsoca.an.2018-04-15T00:00:00Z.nc
```

File: test/testinput/checkpointmodel.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  ice_filename: ice.3dvar.an.2018-04-15T00:00:00Z.nc
```
Comment (travissluka) [inline test/testinput/checkpointmodel.yml]:
```suggestion
  ice_filename: ice.3dvarsoca.an.2018-04-15T00:00:00Z.nc
```

**Review by travissluka — CHANGES_REQUESTED** [review]

File: test/testinput/checkpointmodel.yml (hunk 9 above)
Comment (travissluka) [reply test/testinput/checkpointmodel.yml]: you should remove your test directory every time before a final make/ctest/push

Comment (guillaumevernieres) [reply test/testinput/checkpointmodel.yml]: yes, I should ...

**Review by travissluka — APPROVED** [review]

**Pull request merged**